fvm/evm/types/proof.go

```diff
@@ -0,0 +1,154 @@
+)
+
+type FlowAddress flow.Address
+
+var FlowAddressCadenceType = cadence.TheAddressType
+var FlowAddressSemaType = sema.TheAddressType
+
+func (addr FlowAddress) ToCadenceValue() cadence.Address {
+	return cadence.Address(addr)
+}
+
+type PublicPath string
+
+var PublicPathCadenceType = cadence.ThePathType
+var PublicPathSemaType = sema.PathType
+
+func (p PublicPath) ToCadenceValue() cadence.Path {
+	return cadence.Path{
+		Domain:     common.PathDomainPublic,
+		Identifier: string(p),
+	}
+}
+
+type SignedData []byte
+
+var SignedDataCadenceType = cadence.NewVariableSizedArrayType(cadence.TheUInt8Type)
+var SignedDataSemaType = sema.ByteArrayType
+
+func (sd SignedData) ToCadenceValue() cadence.Array {
+	values := make([]cadence.Value, len(sd))
+	for i, v := range sd {
+		values[i] = cadence.NewUInt8(v)
+	}
+	return cadence.NewArray(values).WithType(SignedDataCadenceType)
+}
+
+type KeyIndices []uint64
+
+var KeyIndicesCadenceType = cadence.NewVariableSizedArrayType(cadence.TheUInt64Type)
+var KeyIndicesSemaType = &sema.VariableSizedType{Type: sema.UInt64Type}
+
+func (ki KeyIndices) ToCadenceValue() cadence.Array {
+	values := make([]cadence.Value, len(ki))
+	for i, v := range ki {
+		values[i] = cadence.NewUInt64(v)
+	}
+	return cadence.NewArray(values).WithType(KeyIndicesCadenceType)
+}
+
+func (ki KeyIndices) Count() int {
+	return len(ki)
+}
+
+type Signature []byte
+
+var SignatureCadenceType = cadence.NewVariableSizedArrayType(cadence.TheUInt8Type)
+
+func (s Signature) ToCadenceValue() cadence.Array {
+	values := make([]cadence.Value, len(s))
+	for i, v := range s {
+		values[i] = cadence.NewUInt8(v)
+	}
+	return cadence.NewArray(values).WithType(SignatureCadenceType)
+}
+
+type Signatures []Signature
+
+var SignaturesCadenceType = cadence.NewVariableSizedArrayType(SignatureCadenceType)
+var SignaturesSemaType = sema.ByteArrayArrayType
+
+func (ss Signatures) ToCadenceValue() cadence.Array {
+	values := make([]cadence.Value, len(ss))
+	for i, s := range ss {
+		values[i] = s.ToCadenceValue()
+	}
+	return cadence.NewArray(values).WithType(SignaturesCadenceType)
+}
+
+func (ss Signatures) Count() int {
+	return len(ss)
+}
+
+type COAOwnershipProofInContext struct {
+	COAOwnershipProof
+	SignedData SignedData
+	EVMAddress Address
+}
+
+func NewCOAOwnershipProofInContext(sd []byte, addr Address, encodedProof []byte) (*COAOwnershipProofInContext, error) {
+	proof, err := COAOwnershipProofFromEncoded(encodedProof)
+	if err != nil {
+		return nil, err
+	}
+	return &COAOwnershipProofInContext{
+		COAOwnershipProof: *proof,
+		SignedData:        sd,
+		EVMAddress:        addr,
+	}, nil
+}
+
+func (proof *COAOwnershipProofInContext) ToCadenceValues() []cadence.Value {
+	return []cadence.Value{
+		proof.Address.ToCadenceValue(),
+		proof.CapabilityPath.ToCadenceValue(),
+		proof.SignedData.ToCadenceValue(),
+		proof.KeyIndices.ToCadenceValue(),
+		proof.Signatures.ToCadenceValue(),
+		proof.EVMAddress.ToCadenceValue(),
+	}
+}
+
```

Comment:
can you add a comment explaining what this proof even is and what the values represent

Reply:
added go docs to the structures.